soroban-env-host/src/storage.rs

```diff
@@ -306,19 +307,55 @@ impl Storage {
         }
     }
 
-    /// Marks the key as a read if we expect the host consumer to read this key.
+    /// Bumps `key` to live for at least `bump_by_ledgers` from now (not
+    /// counting the current ledger).
     ///
-    /// In [FootprintMode::Recording] mode, records the access.
+    /// This operation is only defined within a host as it relies on ledger
+    /// state.
     ///
-    /// In [FootprintMode::Enforcing] mode, succeeds only if the access has been
-    /// declared in the [Footprint].
-    pub fn touch_key(&mut self, key: &Rc<LedgerKey>, budget: &Budget) -> Result<(), HostError> {
-        let _span = tracy_span!("touch key");
-        let ty = AccessType::ReadOnly;
-        match self.mode {
-            FootprintMode::Recording(_) => self.footprint.record_access(key, ty, budget),
-            FootprintMode::Enforcing => self.footprint.enforce_access(key, ty, budget),
+    /// This is always considered a 'read-only' operation, even though it might
+    /// modify the internal storage state.
+    pub fn bump(
+        &mut self,
+        host: &Host,
+        key: Rc<LedgerKey>,
+        bump_by_ledgers: u32,
+    ) -> Result<(), HostError> {
+        let _span = tracy_span!("bump key");
+        let new_expiration =
+            host.with_ledger_info(|li| Ok(li.sequence_number.saturating_add(bump_by_ledgers)))?;
+        // Bumping deleted/non-existing/out-of-footprint entries will result in
+        // an error.
+        let old_entry = self.get(&key, host.budget_ref())?;
+        let old_expiration = match &old_entry.data {
+            LedgerEntryData::ContractData(d) => d.expiration_ledger_seq,
+            LedgerEntryData::ContractCode(c) => c.expiration_ledger_seq,
+            _ => {
+                return Err(host.err(
+                    ScErrorType::Storage,
+                    ScErrorCode::InternalError,
+                    "trying to bump unexpected ledger entry type",
+                    &[],
+                ));
+            }
+        };
+
+        if new_expiration > old_expiration {
+            let mut new_entry = (*old_entry).metered_clone(host.budget_ref())?;
+            match &mut new_entry.data {
+                LedgerEntryData::ContractData(data) => {
+                    data.expiration_ledger_seq = new_expiration;
```

Comment:
This entry will not be in the readwrite set right? How are you going to reconcile parallel updates? That was the reason we initially went with the bump value approach. (cc @graydon)

Reply:
Yeah the point of these being modeled separately is that they are blind writes -- we commute them / sum them together, without any relationship to the footprint

Reply:
There are no parallel updates from the host perspective. The adapter layer between core and host will present the RO updates in a reasonable fashion (so instead of modified entries, core will get modified entries + RO rent bumps).

As I've mentioned in the issue description, the logic was broken in several ways. If we do something like bump(1 year); delete(); create(); bump(1 week); the expectation should be that the entry will end up bumped by 1 week, not by 1 year. Hence this has to be tied to the storage.

Even though the core will end up having some merge logic for extension entries going forward, from the tx standpoint the bumps will be relative to the ledger snapshot, otherwise we run into non-deterministic fees. So the logic to actually merge bumps should live at a higher level than tx application logic (and happen after that).

Reply:
https://github.com/stellar/rs-soroban-env/pull/964 (draft) uses this PR as a base for the shared implementation of ledger change tracking.